### ISSUE TITLE
Cannot call require from thread when I already required autoload.rb

### ISSUE THREAD

**Issue by daniel-rikowski** [rails/rails#22329]
**Disclaimer:** There are lots of autoload and/or thread related issues. I apologize if I missed a duplicate.

I cannot call `require` from a thread when I already required `autoload.rb`

**Example:**
```ruby
require 'active_support/dependencies/autoload'
thread = Thread.new { require 'pathname' }
thread.join
```

* I get an error with `join`:  "No live threads left. Deadlock?"
* Replacing `thread.join` with `sleep` does not result in an exception, but the `require` call inside the thread never returns.
* Removing the first `require` statement makes the code pass.

My test environment:

- Ruby 2.2.3
- ActiveSupport 4.2.5
- Windows 10 and Fedora 23

Is there something I can do to make this work, besides ditching the autoload feature?

**Comment by pixeltrix** [comment]
@daniel-rikowski it's generally regarded as a bad idea to use `require` from a thread - this is why we preload the app in production by loading all of `app` before starting to process requests. Any reason you can't move the require outside of the thread?

**Comment by daniel-rikowski** [comment]
I'm using ActiveSupport outside of Rails, from a GTK application to be precise. 
To keep my user interface responsive I'm performing blocking operations in threads. 

I could move all my own requires outside and call `eager_load!`, but I can't control which 3rd party gems might call `require` from their code. `open-uri` for example performs late requires.

**Comment by matthewd** [comment]
@daniel-rikowski I stuck the above code into the template from the [contribution guide](http://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#reporting-an-issue), and it seems to be working for me: https://gist.github.com/matthewd/e3839872f05cbe9958e3

**Comment by daniel-rikowski** [comment]
**tl;dr**: It's a bug in rubgems in Ruby 2.2. Using 2.1 or 2.3 or bundler fixes the problem.

@matthewd: I ran your code and I didn't get that error, too. That made me very curious and I started to strip down the file and then even the bundler code. Until I found these lines:

```ruby
::Kernel.class_eval do
  if private_method_defined?(:gem_original_require)
    alias rubygems_require require
    alias require gem_original_require
  end

  undef gem
end
```

They seem to make the difference. 

It looks like bundler circumvents a problem with the rubygems require. Switching to a different Ruby version (or using bundler) fixes the problem.

Thank you for your input!
